File: packages/engine/src/cpus/6502/instructions/asl.test.js
import { test, expect } from 'vitest'
import { BufferMapper } from '../../../buffer_mapper'
import { CPU } from '..'

test('asl acc', () => {
  const mmu = new BufferMapper()
  const cpu = new CPU(mmu)
  cpu.acc = 1
  mmu.data = [0x0a] // ASL
  const g = cpu.run()
  for (let i = 0; i < 2; i++) g.next()
  expect(cpu.acc).toBe(2)
})

test('asl zpg', () => {
  const mmu = new BufferMapper()
  const cpu = new CPU(mmu)
  mmu.data = [0x06, 0x02, 4] // ASL $02
  const g = cpu.run()
  for (let i = 0; i < 5; i++) g.next()
  expect(mmu.data[2]).toBe(8)
})

test('asl abs', () => {
  const mmu = new BufferMapper()
  const cpu = new CPU(mmu)
  mmu.data = [0x0e, 0x02, 0xff] // ASL $ff02
  mmu.data[0xff02] = 4
  const g = cpu.run()
  for (let i = 0; i < 6; i++) g.next()
  expect(mmu.data[0xff02]).toBe(8)
})

File: packages/engine/src/cores/famibob/cpu/instructions/asl.test.zig
const testing = @import("../testing/setup.zig");

test "ASL acc" {
    var cpu = testing.createCPU();
    cpu.acc = 1;
    cpu.write(0, 0x0a);
    for (0..2) |_| {
        cpu.process();
    }
    try testing.expectEqual(2, cpu.acc);
}

test "ASL zpg" {
    var cpu = testing.createCPU();
    cpu.write(0, 0x06);
    cpu.write(1, 2);
    cpu.write(2, 4);
    for (0..5) |_| {
        cpu.process();
    }
    try testing.expectEqual(8, cpu.read(2));
}

test "ASL abs" {
    var cpu = testing.createCPU();
    cpu.write(0, 0x0e);
    cpu.write(1, 2);
    cpu.write(2, 0xff);
    cpu.write(0xff02, 4);
    for (0..6) |_| {
        cpu.process();
    }
    try testing.expectEqual(8, cpu.read(0xff02));
}
